zeroex/orderwatch/decoder/event_decoder.go

```diff
@@ -874,7 +874,7 @@ func (d *Decoder) decodeERC721(log types.Log, decodedLog interface{}) error {
 
 	erc721Err := unpackLog(decodedLog, eventName, log, d.erc721ABI)
 	if erc721Err != nil {
-		if _, ok := erc721Err.(UnsupportedEventError); ok {
+		if strings.Contains(erc721Err.Error(), "Unpack(no-values unmarshalled") {
```

Comment:
What's the reason for this change? It functionally changes how we handle `UnsupportedEventError` and looks like it introduces a bug.

Reply:
This change was made due to a change in the way that Go's `abi` package unmarshals events. There is no longer a way to distinguish invalid events by looking at the difference in length between `indexed` and `len(topics) - 1` from what I could tell by trying to fix the bug that was introduced by upgrading. 

The error that is now emitted when an invalid ERC721 asset data is passed to `unpackLog` is something that begins with the error seen above. 

FWIW, this passes all of the tests with the same behavior as the old `if` statement, but I'll grant you that the tests passing is not the same as a bug not existing.

Reply:
Can you move the `strings.Contains(erc721Err.Error(), "Unpack(no-values unmarshalled")` to `unpackLog` instead of doing it here?

Basically change this:

```
if len(indexed) != len(log.Topics[1:]) {
	return UnsupportedEventError{Topics: log.Topics, ContractAddress: log.Address}
}
```

To:

```
if err := _abi.Unpack(decodedEvent, event, log.Data); err != nil {
	if strings.Contains(err.Error(), "Unpack(no-values unmarshalled") {
		return UnsupportedEventError{Topics: log.Topics, ContractAddress: log.Address}
	}
	return err
}
```

We should avoid error string comparisons altogether if possible, however, we can't avoid it here because of [how the error is created in `go-ethereum`](https://github.com/ethereum/go-ethereum/blob/f3f1e59eea922e5fb0d78d55aba471c35c20b022/accounts/abi/argument.go#L95). In this case, we should try and keep the string comparison as close as possible to the original call site (i.e. in this case where we call `ABI.Unpack`).

This will also make the code more consistent with the previous behavior. As it is currently written there are two differences that could cause problems:

1. There are situations where we used to return `UnsupportedEventError` from `unpackLog` but we currently do not. Namely, if `ABI.Unpack` fails.
2. There are situations where we used to return `UnsupportedEventError` from `decodeERC721` but  we currently do not. Namely, if the fallback encoding fails we return `erc721Err`.